omegaconf/dictconfig.py

```diff
@@ -107,24 +107,51 @@ def __init__(
             format_and_raise(node=None, key=None, value=None, cause=ex, msg=str(ex))
 
     def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "DictConfig":
-        res = DictConfig({})
-        for k, v in self.__dict__.items():
-            res.__dict__[k] = copy.deepcopy(v, memo=memo)
-        res._re_parent()
+        res = DictConfig(None)
+        res.__dict__["_metadata"] = copy.deepcopy(self.__dict__["_metadata"], memo=memo)
+        res.__dict__["_flags_cache"] = copy.deepcopy(
+            self.__dict__["_flags_cache"], memo=memo
+        )
+
+        src_content = self.__dict__["_content"]
+        if isinstance(src_content, dict):
+            content_copy = {}
+            for k, v in src_content.items():
+                try:
+                    old_parent = v.__dict__["_parent"]
+                    v.__dict__["_parent"] = None
+                    vc = copy.deepcopy(v, memo=memo)
+                    vc.__dict__["_parent"] = res
+                    content_copy[k] = vc
+                finally:
+                    v.__dict__["_parent"] = old_parent
+        else:
+            # None and strings can be assigned as is
+            content_copy = src_content
+
+        res.__dict__["_content"] = content_copy
+
         return res
 
     def __copy__(self) -> "DictConfig":
-        from .omegaconf import MISSING
-
         res = DictConfig(content=None)
         for k, v in self.__dict__.items():
-            res.__dict__[k] = copy.copy(v)
-        res._re_parent()
+            if k != "_content":
+                res.__dict__[k] = copy.copy(v)
         content = self.__dict__["_content"]
-        if content != MISSING and content is not None:
+        if isinstance(content, dict):
+            content_copy = {}
             for k, v in content.items():
                 if isinstance(v, ValueNode):
-                    res.__dict__["_content"][k] = copy.copy(v)
+                    vc = copy.copy(v)
+                    vc._set_parent(res)
```

Comment:
I don't think this is fixable without making copy actually do a deep copy.
A node can only have one parent. If we copy that parent without copying the node, it will have the wrong parent.

Reply:
If you copy the node, the problem moves to the next level.

```
a <-> b <-> c

# copy just a: b does not link to a1
a <-> b <-> c
a1 -> b <-> c

# copy b as well, now c does not point to b1.
a <->   b  <-> c
a1 <-> b1  -> c
```

Reply:
Yes I agree, and actually I feel like this may be the best solution (make copy behave as deepcopy). Unless you tell me that it's actually ok for a child to point to a different parent, but I'd be suspicious of such a claim (and I expect that trying to support multiple parents is going to open a box that better stay closed).

When you think about it, it would actually be consistent with the current assignment behavior. For instance, one can see the  shallow copy of a dict in Python as doing the following sequence of operations:

```python
def copy_dict(d: dict) -> dict:
   cp = {}
   for k, v in d.items():
        cp[k] = v
   return cp
```

Translated in DictConfigs, this becomes:

```python
def copy_dictconfig(d: DictConfig) -> DictConfig:
    cp = DictConfig({})
    for k, v in d.items():
        cp[k] = v
    return cp
```

And with the current codebase, `copy_dictconfig` is actually returning the equivalent of a deep copy.
(just to be clear, I'm not saying we should implement this function, this is just to say that having `copy.copy(some_dictconfig)` return a shallow copy is consistent with how config assignments work))

Reply:
There is a good reason for DictConfig assignment to do deepcopy (by default), and that is that we would have to destroy the input config otherwise (by changing the parent fields).

I think I agree with you though about making copy do deepcopy.